client.go

```diff
@@ -1,48 +1,273 @@
+// New creates a new S3 Client configured for the given region and bucket name.
+// If HasUserDefinedPSK is true, it will also have a crypto client.
+func New(region string, bucketName string, HasUserDefinedPSK bool) (*S3, error) {
 
-	// AWS credentials gathered from the env.
-	auth, err := aws.GetAuth("", "", "", time.Time{})
+	// Create AWS session with the provided region
+	sess, err := session.NewSession(&aws.Config{Region: &region})
 	if err != nil {
 		return nil, err
 	}
 
-	// Initialise amazon client with internal client.
-	amzCli := &AmzClientImpl{
-		goamzs3.New(auth, aws.Regions[region]),
+	// Create AWS-SDK-S3 client with the session
+	sdkClient := s3.New(sess)
+
+	// If we require crypto client (HasUserDefinedPSK), create it.
+	var cryptoClient S3CryptoClient
+	if HasUserDefinedPSK {
+		cryptoConfig := &s3crypto.Config{HasUserDefinedPSK: true}
+		s3cryptoClient := s3crypto.New(sess, cryptoConfig)
+
+		cryptoClient = s3cryptoClient
 	}
 
-	// Create S3 with the created amzClient.
-	return NewWithClient(amzCli, bucketName), nil
+	cli := Instantiate(sdkClient, cryptoClient, bucketName, region)
+	return cli, nil
 }
 
-// NewWithClient returns a new S3 structure for the provided AmzClient instance.
-func NewWithClient(client AmzClient, bucketName string) *S3 {
+// Instantiate creates a new instance of S3 struct with the provided clients, bucket and region
+func Instantiate(sdkClient S3SDKClient, cryptoClient S3CryptoClient, bucketName, region string) *S3 {
 	return &S3{
-		cli:        client,
-		BucketName: bucketName,
+		sdkClient:    sdkClient,
+		cryptoClient: cryptoClient,
+		bucketName:   bucketName,
+		region:       region,
 	}
 }
 
-// Get returns an io.ReadCloser instance for the given fully qualified S3 URL.
-// Note that this function will mutate the bucket name of the S3 object if it's different than the existing one.
-func (s3 *S3) Get(rawURL string) (io.ReadCloser, error) {
+// BucketName is a getter for the bucket name used by this S3 client
+func (cli *S3) BucketName() string {
+	return cli.bucketName
+}
+
+// Upload handles the uploading a file to AWS S3, into the bucket configured for this client
+func (cli *S3) Upload(ctx context.Context, req *UploadRequest, payload []byte) error {
+	return cli.UploadWithPsk(ctx, req, payload, nil)
+}
+
+// UploadWithPsk handles the uploading a file to AWS S3, into the bucket configured for this client, using a user-defined psk
+func (cli *S3) UploadWithPsk(ctx context.Context, req *UploadRequest, payload []byte, psk []byte) error {
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		log.Event(ctx, "error fetching multipart list", log.Error(err))
+		return err
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
```

Comment:
Does this need a `break` here once the uploadID is assigned?

Reply:
Now it's a return statement, as I created a function to perform the get-or-create atomically